Sources/ApodiniObserve/ConfiguredLogger.swift

```diff
@@ -0,0 +1,232 @@
+//
+// This source file is part of the Apodini open source project
+//
+// SPDX-FileCopyrightText: 2019-2021 Paul Schmiedmayer and the Apodini project authors (see CONTRIBUTORS.md) <paul.schmiedmayer@tum.de>
+//
+// SPDX-License-Identifier: MIT
+// 
+
+import Foundation
+import Logging
+import Apodini
+import ApodiniExtension
+import ApodiniHTTPProtocol
+
+@propertyWrapper
+/// A ``DynamicProperty`` that allows logging of an associated handler
+/// Can be configured with a certain logLevel, so what messages should actually be logged and which shouldn't
+/// Automatically attaches metadata from the handler to the built logger so the developer gets easy insights into the system
+public struct ConfiguredLogger: DynamicProperty {
+    /// The ``Connection`` of the associated handler
+    /// The actual ``Request`` resides here
+    @Environment(\.connection)
+    var connection: Connection
+    
+    /// The ``Storage`` of the ``Application``
+    @Environment(\.storage)
+    var storage: Storage
+    
+    /// The ``Logger`` of the ``Application``
+    @Environment(\.logger)
+    var logger: Logger
+    
+    /// Metadata from the ``Blackboard`` that is injected into the environment of the ``Handler``via a ``Delegate``
+    @Environment(\LoggerExporter.BlackboardMetadata.value)
+    var blackboardMetadata
```

Comment:
CC: @theMomax 

Just a though. This environment property requires that in every case the user properly configured the LoggerExporter in their web service.
If the user forgot to do that this would crash a runtime with some generic error that the keyPath couldn't be found in the environment.
I would propose to provide a proper error message to the user, notifying them that they forgot to configure a logger in their web service (e.g. particularly useful when using off-the-shelve components).
There is straightforward way to check existence of an environment keyPath. In my Authorization PR I did it by declaring the keyPath property as an extension to the `Application`. In those cases, the `@Environment` property will call the getter of the property on the Application if it doesn't find an according value in the environment.
There you could fatal out with a proper runtime error message, maybe?

@theMomax not sure if we should maybe think about a more straightforward way of dynamically checking existence of an environment value(?). This is now the second time that we have this use case I think.

Side note: Am I right with the following: as soon as you use a `@ConfiguredLogger` you are required to have an according LoggerConfiguration in your web service right? Thinking of the use case of using off-the-shelve/third-party components, which uses a `@ConfiguredLogger` (e.g. to provide maximum compatibility), would it make sense to somehow provide a default logger, such that users which don't necessarily have a dedicated Logger configuration can still use such prebuilt components?

Reply:
Good point! I think there are two possible strategies:
* We make `Optional` `Traversable` (in SwiftUI it is). This way a standard component that wants to support a logger could just declare it optional and take an instance in its initializer or use nil as a default. This is currently also possible with `Properties`, but way more ugly. With this approach the service developer is in full control and everything is explicit.
* We promote the path you chose for Authorization by creating such a `_myEnvironment.isPresent`. This would result in more automatism. This could also result in misleading/unexpected scenarios. An additional disadvantage here is that we promote a culture were libraries just setup a default for everything, even though that might never be used. This in turn could lead to bad startup-times in the long term.

I'd suggest to go for the first option, even though that might result in a couple of more lines to be changed when e.g. adding a logger. However, I think we should discuss that in a bigger group! The good thing is neither of the options are difficult to implement.

Reply:
> This in turn could lead to bad startup-times in the long term.

I don't quite understand that argument. But I would probably in any case advocate for option 1. Seems more straight forward and natural to me. I think this would also work quite well for the `Authorized` property and the `@ConfiguredLogger` propertyWrapper. Both of those can declare their `@Environment` optional and provide according failure logic in those events.

I don't want to block Philipps work though. Should the result of those discussions be implemented in a separate PR? 

Whats your take on my second point, about having some sort of default logger in the case where no Logger configuration is present. Does that even make sense? I mean the propertyWrapper is called `@ConfiguredLogger` and not just `@Logger`. Would it maybe make sense to create a `@Logger` which falls back to the default Application logger if nothing is configured? Maybe I'm missing some of the vision behind the ApodiniObserve package?

Reply:
> I don't want to block Philipps work though. Should the result of those discussions be implemented in a separate PR?

Yes, definitely!

> Whats your take on my second point, about having some sort of default logger in the case where no Logger configuration is present. Does that even make sense? I mean the propertyWrapper is called @ConfiguredLogger and not just @Logger. Would it maybe make sense to create a @Logger which falls back to the default Application logger if nothing is configured? Maybe I'm missing some of the vision behind the ApodiniObserve package?

I think the `Configured` prefix is just to avoid naming collision with the normal `Logger` type. (Maybe `ApodiniLogger` or `AutoLogger` would be a better fit, though.) I don't think a default logger configuration makes sense. With the solution to the first problem (making `Optional` `Traversable`), we don't need a default logger, because that enables standard components to just not use a logger if not explicitly enabled by the service developer. We could add a default logger based on a `canImport()`, but I don't really see the point in adding such magic, as the consequence of not having a default logger is basically the following:
1. Service Developer forgets to add a logger configuration but uses `@ConfiguredLogger` somewhere in the code.
2. They get a fatal error while testing their change
3. They do a quick online search
4. They add the logger configuration item and gained a better understanding of the framework from the beginning

Idk...I just dislike frameworks that do everything automatically because I have no clue what is going on until I need to configure something. And when I need to do so, I first have to search for where I can even add the configuration, because the basic configuration wasn't part of the getting started guide - because the default configuration is just added somewhere behind the curtain if not explicitly stated otherwise.

Reply:
I agree with your points made. Specifically summarized with:

> With the solution to the first problem (making Optional Traversable), we don't need a default logger, because that enables standard components to just not use a logger if not explicitly enabled by the service developer.

Maybe as a note to @philippzagar. Should we incorporate that ability to optionally call I logger, I would love to see the logging framework be incorporated into ApodiniAuthorization. Specifically in the `Authenticator` and the `AuthorizationRequirementsChecker`.
Those two handlers log the result of the AuthorizationRequirements evaluation wich might be important for debugging reasons, to be able to track why a given request was rejected.

E.g. logging statements in 
https://github.com/Apodini/Apodini/blob/251bbba60b1023443b6973d80ba42c4155fe143a/Sources/ApodiniAuthorization/Delegates/Authenticator.swift#L120

or  

https://github.com/Apodini/Apodini/blob/251bbba60b1023443b6973d80ba42c4155fe143a/Sources/ApodiniAuthorization/Delegates/AuthorizationRequirementsChecker.swift#L59

Reply:
So, I'll revolve this conversation for now and we will pick it up again sometime after this PR is merged and both of you have submitted your thesis :)